### PR TITLE
[rewarding] move claim address check to LoadProto() and SanityCheck()

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
#4277 added a `claimTo` field to receive claim reward, so other EOA can help claim reward on another's behalf.

move the address check from inside  `protocol.Handle` to the action's `LoadProto()` and `SanityCheck()`, which is consistent with other action's behavior

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/rewarding/reward.go [hunk 1]
```diff
@@ -365,10 +365,10 @@ func (p *Protocol) grantToAccount(ctx context.Context, sm protocol.StateManager,
 	return p.putState(ctx, sm, accKey, &acc)
 }
 
-func (p *Protocol) claimFromAccount(ctx context.Context, sm protocol.StateManager, addr address.Address, amount *big.Int) error {
+func (p *Protocol) claimFromAccount(ctx context.Context, sm protocol.StateManager, from, to address.Address, amount *big.Int) error {
 	// Update reward account
 	acc := rewardAccount{}
-	accKey := append(_adminKey, addr.Bytes()...)
+	accKey := append(_adminKey, from.Bytes()...)
```
Comment (dustinxie) [inline action/protocol/rewarding/reward.go:371]: the current code used `claimTo` from L288, there's no balance in this address (if different from `actionCtx.Caller`)

**Comment by CoderZhi** [comment]
#4277 added a new field `address` to action `ClaimFromRewardingFund`, and the purpose is to help other accounts to claim rewards (not claiming rewards to some other accounts).
`claimTo` is just a parameter name, which may be confusing. a better naming for this parameter is `claimFrom`.

File: action/protocol/rewarding/protocol.go [hunk 2]
```diff
@@ -215,16 +215,11 @@ func (p *Protocol) Handle(
...
-			claimTo = protocol.MustGetActionCtx(ctx).Caller.String()
+		addr := act.ClaimAccount()
+		if addr == nil {
+			addr = protocol.MustGetActionCtx(ctx).Caller
 		}
-		claimToAddr, err := address.FromString(claimTo)
-		if err != nil {
-			log.L().Debug("Invalid claim to address", zap.Error(err))
-			return p.settleUserAction(ctx, sm, uint64(iotextypes.ReceiptStatus_Failure), si, nil)
-		}
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol.go:226]: this address check can (and should) be done before `Handle()` in the action's `LoadProto()` and `SanityCheck()`, this is consistent with other actions like candidate register

File: action/protocol/rewarding/protocol.go [hunk 3]
```diff
@@ -215,16 +215,11 @@ func (p *Protocol) Handle(
 		return p.settleUserAction(ctx, sm, uint64(iotextypes.ReceiptStatus_Success), si, nil, rlog)
 	case *action.ClaimFromRewardingFund:
 		si := sm.Snapshot()
-		claimTo := act.Address()
-		if claimTo == "" {
-			claimTo = protocol.MustGetActionCtx(ctx).Caller.String()
+		addr := protocol.MustGetActionCtx(ctx).Caller
+		if protocol.MustGetFeatureCtx(ctx).AddClaimRewardAddress && act.Address() != nil {
```
Comment (CoderZhi) [inline action/protocol/rewarding/protocol.go:219]: shall we reject an action if `!protocol.MustGetFeatureCtx(ctx).AddClaimRewardAddress && act.Address() != nil`?

**Review by CoderZhi — COMMENTED** [review]
lgtm, except the comment below

File: action/protocol/rewarding/protocol.go (hunk 3 above)
Comment (envestcc) [reply action/protocol/rewarding/protocol.go:219]: can check it when validating

File: action/claimreward.go [hunk 4]
```diff
@@ -217,8 +232,16 @@ func NewClaimFromRewardingFundFromABIBinary(data []byte) (*ClaimFromRewardingFun
 	if ac.data, ok = paramsMap["data"].([]byte); !ok {
 		return nil, errDecodeFailure
 	}
-	if ac.address, ok = paramsMap["address"].(string); !ok {
+	var s string
+	if s, ok = paramsMap["address"].(string); !ok {
```
Comment (envestcc) [inline action/claimreward.go:236]: should keep backword compatibility, allowing without address param

File: action/protocol/rewarding/protocol.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/rewarding/protocol.go:224]: added

File: action/claimreward.go (hunk 4 above)
Comment (dustinxie) [reply action/claimreward.go:286]: add a 2nd commit to keep both v1 and v2 ABI

File: action/claimreward.go [hunk 5]
```diff
@@ -15,11 +15,30 @@ import (
...
+				"internalType": "uint8[]",
+				"name": "data",
+				"type": "uint8[]"
+			}
+		],
+		"name": "claim",
+		"outputs": [],
+		"stateMutability": "nonpayable",
+		"type": "function"
+	},
```
Comment (dustinxie) [inline action/claimreward.go:41]: need to keep the current (v1) ABI for backward compatibility

File: action/rlp_tx_test.go [hunk 6]
```diff
@@ -299,21 +299,21 @@ var (
...
+			"100",
+			"0",
+			"0xA576C141e5659137ddDa4223d209d4744b2106BE",
+			_evmNetworkID,
+			iotextypes.Encoding_ETHEREUM_EIP155,
+			100,
+			"4d26d0736ebd9e69bd5994f3730b05a2d48c810b3bb54818be65d02004cf4ff4",
+			"04830579b50e01602c2015c24e72fbc48bca1cca1e601b119ca73abe2e0b5bd61fcb7874567e091030d6b644f927445d80e00b3f9ca0c566c21c30615e94c343da",
+			"8d38efe45794d7fceea10b2262c23c12245959db",
+		},
```
Comment (dustinxie) [inline action/rlp_tx_test.go:316]: this is test with v1 ABI, fixed and restored
the original #4277 shouldn't have broken this

File: action/claimreward.go [hunk 7]
```diff
@@ -185,15 +209,23 @@ func (b *ClaimFromRewardingFundBuilder) Build() ClaimFromRewardingFund {
...
+	if c.address == nil {
+		// this is v1 ABI before adding address field
+		data, err := _claimRewardingMethodV1.Inputs.Pack(c.Amount(), c.Data())
+		if err != nil {
+			return nil, err
+		}
+		return append(_claimRewardingMethodV1.ID, data...), nil
+	}
 	var addr string
 	if c.address != nil {
```
Comment (CoderZhi) [inline action/claimreward.go:221]: delete

File: action/claimreward.go [hunk 8]
```diff
@@ -214,16 +246,34 @@ func (c *ClaimFromRewardingFund) ToEthTx(_ uint32) (*types.Transaction, error) {
...
+		act *ClaimFromRewardingFund
+		err error
+	)
+	if act, err = newClaimFromRewardingFundFromABIv2(data); err == nil {
+		return act, nil
+	}
+	if err != errWrongMethodSig {
+		// data is v2-encoded, but got error when decoding
+		return nil, err
+	}
```
Comment (CoderZhi) [inline action/claimreward.go:259]:
```
    if len(data) >= 4 {
        switch {
        case bytes.Equal(_claimRewardingMethodV2.ID[:], data[:4]):
              ....
        case bytes.Equal(_claimRewardingMethodV1.ID[:], data[:4]):
              ....
        }
    }
    return nil, errDecodeFailure
```

**Review by CoderZhi — COMMENTED** [review]
reverse order if doesn't hurt, because `data` is useless for now

**Review by CoderZhi — APPROVED** [review]

**Review by envestcc — APPROVED** [review]

File: action/claimreward.go (hunk 7 above)
Comment (dustinxie) [reply action/claimreward.go:221]: fixed

**Pull request merged**